EIPS/draft_token_validation.md

```diff
@@ -0,0 +1,181 @@
+
+> parameters
+> * `token`: the token under review
+> * `user`: the user to check
+>
+> *returns* an 8-bit status code
+
+##### check/4
+
+`function check(address token, address from, address to, uint256 amount) public returns (uint8 resultCode)`
+
+> parameters
+> * `token`: the token under review
+> * `from`: in the case of a transfer, who is relinquishing token ownership
+> * `to`: in the case of a transfer, who is accepting token ownership
+> * `amount`: The number of tokens being transferred
+>
+> *returns* an 8-bit status code
+
+### ValidatedToken
+
+```solidity
+interface ValidatedToken {
+  event Validation(
+    address indexed user,
+    uint8   indexed result
+  )
+
+  event Validation(
+    address indexed from,
+    address indexed to,
+    uint256 value,
+    uint8   indexed result
+  )
+}
+```
+
+#### Events
+
+##### Validation/2
+
+`event Validation(address indexed user, uint8 indexed resultCode)`
+
+This event MUST be fired on return from a call to a `TokenValidator.check/2`.
+
+> parameters
+> * `user`: the user to check
+> * `resultCode`: a status code
+
+
+##### Validation/4
+
+```solidity
+event Validation(
+  address indexed from,
+  address indexed to,
+  uint256 amount,
+  uint8   indexed result
+)
+```
+
+This event MUST be fired on return from a call to a `TokenValidator.check/4`.
+
+> parameters
+> * `from`: in the case of a transfer, who is relinquishing token ownership
+> * `to`: in the case of a transfer, who is accepting token ownership
+> * `amount`: The number of tokens being transferred
+> * `resultCode`: a status code
+
+## Rationale
+
+This proposal includes a financial permissions system on top of any financial token.
+This design is not a general roles/permission system. In any system, the more you know
+about the context where a function will be called, the more powerful your function can be.
+By restricting ourselves to token transfers (ex. ERC20 or EIP-777), we can make
+assumptions about the use cases our validators will need to handle, and can make
+the API both small, useful, and extensible.
+
+The events are fired by the calling token. Since `Validator`s may aggregate or delegate
+to other `Validator`s, it would generate a lot of useless events were it the
+`Validator`'s responsibility. This is also the reason why we include the `token`
+in the `call/4` arguments: a `Validator` cannot rely on `msg.sender` to determine
+the token that the call is concerning.
+
+We have also seen a similar design from [R-Token](https://github.com/harborhq/r-token) that uses an additional field: `spender`.
+While there are potential use cases for this, it's not widely used enough to justify passing
+a dummy value along with every call. Instead, such a call would look more like this:
+
+```solidity
+function approve(address spender, uint amount) public returns (bool success) {
+    if(validator.check(this, msg.sender, spender, amount) == !okStatusCode) {
```

Comment:
> `return` is an evil `goto`

I agree in this case. Are you not a fan of guard statements generally, or just here?

Reply:
well, if _truly_ a guard, wouldn't you use `require()` :trollface:  ? 

but yeah, overall, not a big fan when, say, commenting out the return statement would cause the function to execute the `else` - just use `else`.  more explicit, less bug chance, etc.

Reply:
Well, I might not want it to blow up, but yes, indeed way more idiomatic in Solidity 👌🏻